idaes/core/util/scaling.py

```diff
@@ -1041,3 +1047,165 @@ def set_derivative_factor_from_state(self, deriv, nominal_wrt=1.0):
+
+    def _get_scale_factor_for_unary_function(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 1
+        func_name = node.getname()
+        func = getattr(math, func_name)
+        return [func(i) for i in child_scale_factors[0]]
+
+    def _get_scale_factor_expr_if(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 3
+        return child_scale_factors[1] + child_scale_factors[2]
+
+    def _get_scale_factor_no_children(self, node, child_units):
+        assert len(child_units) == 0
+        # may need more checks for dimensionless for other types
+        assert type(node) is IndexTemplate
+        return [1]
+
+    def _get_scale_factor_external_function(self, node, child_units):
+        # TODO: Work out what to do here
+        # Maybe pass child_units into the external function and get the result?
+        # For now, assume we know nothing about scaling
+        return [1]
+
+    node_type_method_map = {
+        EXPR.EqualityExpression: _get_scale_factor_for_sum,
+        EXPR.InequalityExpression: _get_scale_factor_for_sum,
+        EXPR.RangedExpression: _get_scale_factor_for_sum,
+        EXPR.SumExpression: _get_scale_factor_for_sum,
+        EXPR.NPV_SumExpression: _get_scale_factor_for_sum,
+        EXPR.ProductExpression: _get_scale_factor_for_product,
+        EXPR.MonomialTermExpression: _get_scale_factor_for_product,
+        EXPR.NPV_ProductExpression: _get_scale_factor_for_product,
+        EXPR.DivisionExpression: _get_scale_factor_for_division,
+        EXPR.NPV_DivisionExpression: _get_scale_factor_for_division,
+        EXPR.PowExpression: _get_scale_factor_for_power,
+        EXPR.NPV_PowExpression: _get_scale_factor_for_power,
+        EXPR.NegationExpression: _get_scale_factor_single_child,
+        EXPR.NPV_NegationExpression: _get_scale_factor_single_child,
+        EXPR.AbsExpression: _get_scale_factor_single_child,
+        EXPR.NPV_AbsExpression: _get_scale_factor_single_child,
+        EXPR.UnaryFunctionExpression: _get_scale_factor_for_unary_function,
+        EXPR.NPV_UnaryFunctionExpression: _get_scale_factor_for_unary_function,
+        EXPR.Expr_ifExpression: _get_scale_factor_expr_if,
+        EXPR.ExternalFunctionExpression: _get_scale_factor_external_function,
+        EXPR.NPV_ExternalFunctionExpression: _get_scale_factor_external_function,
+        EXPR.LinearExpression: _get_scale_factor_for_sum,
+    }
+
+    def exitNode(self, node, data):
+        """Callback for :class:`pyomo.core.current.StreamBasedExpressionVisitor`. This
+        method is called when moving back up the tree in a depth first search."""
+
+        # first check if the node is a leaf
+        nodetype = type(node)
+
+        # TODO: What do we do with an indexed component? Fail outright?
+
+        if nodetype in native_types or nodetype in pyomo_constant_types:
+            if node == 0:
+                return [1]  # Catch cases of zero and use default scaling of 1
```

Comment:
In line with what I said about summations, positive infinity would be the best scaling factor to assign to zero, because then it wouldn't throw off the harmonic sum of scaling factors (say, if you were summing 0.001 + 0.01 + 0). Not sure whether that would cause problems down the line elsewhere, though.

Reply:
This should not matter in most cases, a Pyomo will eliminate the 0 term anyway - however we need to have a real number as the scaling factor to avoid evaluation errors elsewhere. There is also a potential case of 0 actually being `False`, in which case `1` is probably the best scaling factor.